### PR TITLE
Ticket workflow for adding gears

### PR DESCRIPTION
**Pull request by kofalt**
This adds a ticket-based workflow for proposing and saving gears.
Used in conjunction with our custom image registry.

This adds four new routes, which may seem like a bit much. One is (designed to be) used exclusively by the registry authorizer, and later two old routes can be dropped.

Coming in a next commit will be the last addition to the save endpoint that ties this flow into that. For now, any ticket save will succeed.

### CODE REVIEW

File: api/jobs/handlers.py [hunk 1]
```diff
@@ -73,6 +75,94 @@ def check(self):
         check_for_gear_insertion(self.request.json)
         return None
```
Comment (nagem) [inline api/jobs/handlers.py]: Was the `no-cover` for development purposes? I see no reason we can't have some integration tests around this workflow.

File: api/api.py [hunk 2]
```diff
@@ -161,9 +161,20 @@ def prefix(path, routes):
             route('/<:[^/]+>/accept-failed-output', JobHandler,  h='accept_failed_output', m=['POST']),
         ]),
         route('/gears',                                  GearsHandler),
-        route('/gears/check',                            GearsHandler, h='check', m=['POST']),
-        route('/gears/temp',                             GearHandler, h='upload', m=['POST']),
-        route('/gears/temp/<cid:{oid}>',  GearHandler, h='download', m=['GET']),
+
+        # New upload flow
+        route('/gears/prepare-add',                      GearsHandler, h='prepare_add',     m=['POST']),
```
Comment (ehlertjd) [inline api/api.py]: Consider a more conventional URI scheme?
* `POST /gears/tickets` - Create ticket for gear creation
* `GET /gears/tickets` - Get own tickets 
* `GET /gears/tickets/<id>` - Get single ticket
* `POST /gears/tickets/<id>` - Save gear to database after upload

File: api/jobs/handlers.py [hunk 3]
```diff
@@ -73,6 +75,94 @@ def check(self):
...
+        """
+
+        # Allow for just a summary of gear names
+        gear_names_only = self.is_true('gear_names_only')
+
+        result = config.db.gear_tickets.find({
+            'origin': self.origin
+        })
+
+        if gear_names_only:
```
Comment (ehlertjd) [inline api/jobs/handlers.py]: What's the use case for names only?

File: api/jobs/handlers.py [hunk 4]
```diff
@@ -73,6 +75,94 @@ def check(self):
...
+    @require_admin
+    def save(self, _id): # pragma: no cover
+        """
+        Save a gear described by an upload ticket.
+        """
+
+        submit = self.request.json
+
+        ticket = config.db.gear_tickets.find_one({
+            '_id': bson.ObjectId(submit['ticket'])
```
Comment (ehlertjd) [inline api/jobs/handlers.py]: Raise 400/404 if `_id` is not a valid ObjectId

File: api/web/base.py [hunk 5]
```diff
@@ -417,7 +417,8 @@ def handle_exception(self, exception, debug, return_json=False): # pylint: disab
             code = 500
 
         if code >= 400 and code < 500:
-            self.log.debug('client error: {}'.format(exception))
+            if 'elastic search not available: ConnectionError' not in str(exception):
```
Comment (ehlertjd) [inline api/web/base.py]: Guessing this shouldn't be here?

File: api/jobs/handlers.py [hunk 6]
```diff
@@ -73,6 +75,94 @@ def check(self):
...
+        }
+
+    @require_admin
+    def get_ticket(self, _id): # pragma: no cover
+        """
+        Retrieve a gear-upload ticket.
+        """
+
+        result = config.db.gear_tickets.find_one({
+            '_id': bson.ObjectId(_id)
```
Comment (ehlertjd) [inline api/jobs/handlers.py]: Raise 400/404 if `_id` is not a valid ObjectId

File: api/jobs/handlers.py (hunk 3 above)
Comment (kofalt) [reply api/jobs/handlers.py]: The registry authorization, which doesn't need any other information.

File: api/web/base.py (hunk 5 above)
Comment (kofalt) [reply api/web/base.py]: Hmm, yeah this is something I've had in my local environment forever. This just squelches some ES flak when ES is not around. I'll remove it or move to a different PR

File: api/jobs/handlers.py (hunk 4 above)
Comment (kofalt) [reply api/jobs/handlers.py]: We don't do that elsewhere in the API... are you wanting that to be a new standard? Maybe we should make a `try_object_id` util function.

File: api/jobs/handlers.py (hunk 1 above)
Comment (kofalt) [reply api/jobs/handlers.py]: Duplicated from other sigs in the file. Most of this logic could have tests

File: api/api.py (hunk 2 above)
Comment (kofalt) [reply api/api.py]: Yeah, that's cleaner.

**Pull request closed without merging**